enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,359 @@
+  are not specified on the route.
+
+### Risks and Mitigations
+
+There is a possibility of an invalid route being processed by the router (edge case),
+if any changes are done to the rbac or the referenced secret is deleted after the API
+server validation but before router has processed the route (maybe router pod is not running)
+then this can lead to the router processing this incorrect route.
+
+> Will need to duplicate the validations present on the API server to the router.
+
+### Drawbacks
+
+The user will need to manually create, provide and maintain the rbac required by the
+router so that it can access secrets securely. This becomes tedious when users have
+100s of Routes.
+
+The workaround for this is to document various levels of rbac that can be provided,
+
+- Grant router service account access to secret by secret-name (explicit rbac)
+- Grant router service account access to all secrets in a fixed namespace (implicit rbac)
+
+The above variations need to be documented for the end user as part of OpenShift documentation.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Performance testing of openshift-router in tech-preview? Is there
+  a workflow present where we can gather some early metrics (memory, cpu)? This will help in
+  preemptively addressing performance concerns before going GA.
+
+- Do we make changes to the ingress-to-route controller as well?
+  > The ingress-to-route behaviour will remain as is i.e. it will not make use of
+  > the newly introduced field.
+
+### Test Plan
+
+Update router tests in openshift/origin and supplement all existing certificate related tests
+with new tests utilizing `.spec.tls.externalCertificate`. Ensure the tests cover the following scenarios,
+
+- Updating routes from default certificates to certificate referenced via
+  secrets and vice-versa.
+- Updating secret/certificate referenced in routes and verify serving
+  certificate has been updated.
+- Updating secret/certificate with incorrect information and verify route
+  is not admitted due to validation failure. (eg: mismatched public and private key, etc)
+
+### Graduation Criteria
+
+This feature will initially be released as Tech Preview only. The e2e tests
+in openshift/origin will only be added when graduating this feature to GA.
+
+#### Dev Preview -> Tech Preview
+
+N/A. This feature will go directly to Tech Preview.
+
+#### Tech Preview -> GA (Future work)
+
+The e2e tests as part of openshift/origin should be consistently passing.
+The router will need to undergo performance testing as part of OCP payload
+to ensure the memory implications of creating and maintains all the active watches
+is verified to be efficient.
+
+##### Future work
+
+The ingress-to-route controller in the route-controller-manager will need to
+be updated to ensure that the created routes use `.spec.tls.externalCertificate`
+instead of `.spec.tls.certificate`. Additional tests will need to be added into
+o/origin for this scenario.
```

Comment:
That's more a question for the future but who will do the necessary RBAC in that case?

Reply:
I think we can still rely on the user providing this rbac if we decide to implement this in the future.

Reply:
Is it only me reading this like `.spec.tls.externalCertificate` will be preferred solution over `.spec.tls.certificate` when the former will become GA?

Reply:
That is what I had in mind, but we can take a collective call. Figured in the future when KMS is brought in it is better if we have the certificate in a secret as opposed to in a route.